config/src/main/java/org/springframework/security/config/annotation/web/configurers/AbstractInterceptUrlConfigurer.java

```diff
@@ -98,62 +101,62 @@ public void configure(H http) throws Exception {
+	 * {@link FilterSecurityInterceptor}. Cannot be null.
 	 */
-	private AccessDecisionManager createDefaultAccessDecisionManager(H http) {
-		AffirmativeBased result = new AffirmativeBased(getDecisionVoters(http));
-		return postProcess(result);
-	}
-
-	/**
-	 * If currently null, creates a default {@link AccessDecisionManager} using
-	 * {@link #createDefaultAccessDecisionManager(HttpSecurityBuilder)}. Otherwise returns
-	 * the {@link AccessDecisionManager}.
-	 * @param http the builder to use
-	 * @return the {@link AccessDecisionManager} to use
-	 */
-	private AccessDecisionManager getAccessDecisionManager(H http) {
-		if (this.accessDecisionManager == null) {
-			this.accessDecisionManager = createDefaultAccessDecisionManager(http);
-		}
-		return this.accessDecisionManager;
-	}
+	abstract AuthorizationManager<FilterInvocation> createAuthorizationManager();
 
 	/**
 	 * Creates the {@link FilterSecurityInterceptor}
-	 * @param http the builder to use
 	 * @param metadataSource the {@link FilterInvocationSecurityMetadataSource} to use
 	 * @param authenticationManager the {@link AuthenticationManager} to use
 	 * @return the {@link FilterSecurityInterceptor}
 	 * @throws Exception
 	 */
-	private FilterSecurityInterceptor createFilterSecurityInterceptor(H http,
+	private FilterSecurityInterceptor createFilterSecurityInterceptor(
 			FilterInvocationSecurityMetadataSource metadataSource, AuthenticationManager authenticationManager)
 			throws Exception {
 		FilterSecurityInterceptor securityInterceptor = new FilterSecurityInterceptor();
 		securityInterceptor.setSecurityMetadataSource(metadataSource);
-		securityInterceptor.setAccessDecisionManager(getAccessDecisionManager(http));
 		securityInterceptor.setAuthenticationManager(authenticationManager);
+		if (this.accessDecisionManager != null) {
+			securityInterceptor.setAccessDecisionManager(this.accessDecisionManager);
+		}
+		else {
+			securityInterceptor.setAuthorizationManager(createAuthorizationManager());
+		}
 		securityInterceptor.afterPropertiesSet();
 		return securityInterceptor;
 	}
 
 	public abstract class AbstractInterceptUrlRegistry<R extends AbstractInterceptUrlRegistry<R, T>, T>
 			extends AbstractConfigAttributeRequestMatcherRegistry<T> {
 
+		private final FilterInvocationAuthorizationManager.Builder managerBuilder = FilterInvocationAuthorizationManager
+				.builder();
+
 		AbstractInterceptUrlRegistry() {
 		}
 
+		final void addMapping(Iterable<? extends RequestMatcher> requestMatchers,
+				AuthorizationManager<FilterInvocationAuthorizationContext> authorizationManager) {
+			for (RequestMatcher requestMatcher : requestMatchers) {
+				addMapping(requestMatcher, authorizationManager);
+			}
+			resetUnmappedMatchers();
+		}
+
+		final void addMapping(RequestMatcher requestMatcher,
+				AuthorizationManager<FilterInvocationAuthorizationContext> authorizationManager) {
+			this.managerBuilder.add(new RequestMatcherEntry<>(requestMatcher, authorizationManager));
```

Comment:
One option might be to change `WebExpressionAuthorizationManager`'s `ExpressionHandler` into a setter so that it can be post-processed by `createAuthorizationManager`. `WebExpressionAuthorizationManager` would use `DefaultWebSecurityExpressionHandler` by default, but would be updated by `createAuthorizationManager`.

Then, while there is still parallel registration, at least the user's expression handler configuration doesn't get missed and the authorization manager would get all the rules in the right order.

Another option feels pretty hacky, but it would be to introduce a private class called `AuthorizationManagerConfigAttribute` that holds onto both items. Then this registry could hold onto config attributes and authorization managers in the same map. This has one minor benefit of removing the `resetUnmappedMatchers` method.

Reply:
In the first option, the expression should be parsed only after the `expressionHandler` is set, since the user can use a custom `expressionParser` in the `expressionHandler`, this seems to complicate the initialization of the `WebExpressionAuthorizationManager` and I'd like to avoid using lifecycle hooks such as `afterPropertiesSet` in this case, so I think the second option is better.